Extend packages.yaml to allow custom attributes (#16526)
LLNL/spack: lib/spack/docs/packaging_guide.rst

Making a package discoverable with ``spack external find``
----------------------------------------------------------

The simplest way to make a package discoverable with
:ref:`spack external find <cmd-spack-external-find>` is to:

1. Mark the Package with the ``detectable`` decorator
2. Define the executables associated with the package
3. Implement a method to determine the versions of these executables

^^^^^^^^^^^^^^^^^
Minimal detection
^^^^^^^^^^^^^^^^^

The first two steps are fairly simple, as they require only to
decorate the package being defined and specify a package level
``executables`` attribute:

.. code-block:: python

   @detectable
   class Foo(Package):
       # Each string provided here is treated as a regular expression, and
       # would match for example 'foo', 'foobar', and 'bazfoo'.
       executables = ['foo']

This attribute must be a list of strings. Each string is treated as a regular
expression (e.g. 'gcc' would match 'gcc', 'gcc-8.3', 'my-weird-gcc', etc.) to

> determine a set of system executables that might be part or this package.

Reviewer: It would be good to mention here that if you only want to match a single executable name, you can use `^foo$`, since that's the most common use case.

Author: Done in f54dcbe

Making a package discoverable with ``spack external find``
----------------------------------------------------------

The simplest way to make a package discoverable with
:ref:`spack external find <cmd-spack-external-find>` is to:

1. Mark the Package with the ``detectable`` decorator
2. Define the executables associated with the package
3. Implement a method to determine the versions of these executables

^^^^^^^^^^^^^^^^^
Minimal detection
^^^^^^^^^^^^^^^^^

The first two steps are fairly simple, as they require only to
decorate the package being defined and specify a package level
``executables`` attribute:

.. code-block:: python

   @detectable
   class Foo(Package):
       # Each string provided here is treated as a regular expression, and
       # would match for example 'foo', 'foobar', and 'bazfoo'.
       executables = ['foo']

This attribute must be a list of strings. Each string is treated as a regular
expression (e.g. 'gcc' would match 'gcc', 'gcc-8.3', 'my-weird-gcc', etc.) to
determine a set of system executables that might be part or this package. Note

> that to match only executables named 'gcc' the regular expression ``r'^gcc$'``

Reviewer 2: Can you mention these are regular expressions? Other than that this explanation looks great.

Author: Changed "Each string is treated as a regular expression" to "Each string is a regular expression"

must be used.

Finally to determine the version of each executable the ``determine_version``
method must be implemented:

.. code-block:: python

   @classmethod
   def determine_version(cls, exe):
       """Return either the version of the executable passed as argument
       or ``None`` if the version cannot be determined
       """

This method receives as input the path to a single executable and must return

> as output its version as a string, or ``None`` is the version cannot be determined.

Reviewer 2: One nice feature of the original implementation of `determine_spec_details` was that if it returned `None`, it was assumed that the provided executable wasn’t an instance of the package, does `determine_version` have this property? Or is there some way to say for a provided executable "this isn't actually an instance of the package"?

Perhaps the same outcome could be achieved by raising a special-purpose exception if `determine_version` realizes that the exe isn't actually an instance of the package.: "NotAnInstanceOfPackage".

Note: `filter_detected_exes` could be useful for many cases of this, but I think there may be redundancy between that and `determine_version`, since typically the later will involve invoking the executable (at which point it is generally clear from the output in some cases that it is not an instance of the package).

Author: > One nice feature of the original implementation of determine_spec_details was that if it returned None, it was assumed that the provided executable wasn’t an instance of the package, does determine_version have this property?

Yes, it has the same property (in details if the version is false-ish the executable won't be added to `exes_by_version` and thus won't proceed in the workflow), see:

https://github.com/spack/spack/blob/b52c236a2ca7843042e6049c6e06596aacb2b9aa/lib/spack/spack/package.py#L2238-L2246

> Perhaps the same outcome could be achieved by raising a special-purpose exception if determine_version realizes that the exe isn't actually an instance of the package.: "NotAnInstanceOfPackage".

Not sure this would be an improvement or a complication for users. In general with the current version users just need to return a value. With the modification proposed they need to return a value or raise an appropriate exception. For the time being I would keep the semantics as it is and maybe modify it in a later PR when we have more experience on a broader set of use cases.

Reviewer 2: > Yes, it has the same property (in details if the version is false-ish the executable won't be added to exes_by_version 

Excellent: can you mention that in the docs here?

Author: Done

Implementing the three steps above is mandatory, and gives the package the

> basic ability to detect if a spec is present on the system at a given version.

Reviewer 3: I think it would be nice to link to [the Executable](https://spack.readthedocs.io/en/latest/spack.util.html#spack.util.executable.Executable) eg

```rst
This method receives as input a single :ref:`~spack.util.executable.Executable`
```

I also think that it would be helpful to round out this example since the same code would likely work for many packages, posting an out-of-spack hack test as demo, but i'm not very attached to this (consistency questions, none of the other stubs have sample implementations -- I just worry that this method particularly how to get stdout back is not straightforward for users unfamiliar with subprocess in general):

```py
#!/usr/bin/env python3
import sys
import re  # NOTE: spack magic imports, do they need re?
sys.path.insert(0, "/opt/spack/lib/spack")  # ;)

class Pkg:
    @classmethod
    def determine_version(cls, exe):
        version = exe("--version", output=str)
        match = re.search(r"(\d\.\d\.\d)", version)
        if match:
            return match.groups()[0]

        return None

if __name__ == "__main__":
    from spack.util.executable import which
    gcc = which("gcc")
    print(f"Version: {Pkg.determine_version(gcc)}")
```

Author: I'll refine the docs, but the method receives as input the path to a single executable not an `Executable` object. The rationale for that is that passing the path is more flexible and constructing an `Executable` object is doable in one line when there's need for that.

Author: Now docs state explicitly that the *path to an executable* is passed as input.

try to search for executables that match ``g++``, that would have
the unwanted side effect of selecting also ``clang++`` - which is
a C++ compiler provided by another package - if present on the system.
Trying to select executables that contain ``g++`` but not ``clang``
would be quite complicated to do using regex only. Employing the
``filter_detected_exes`` method it becomes:

.. code-block:: python

   class Gcc(Package):
      executables = ['g++']

      def filter_detected_exes(cls, prefix, exes_in_prefix):
         return [x for x in exes_in_prefix if 'clang' not in x]

Another possibility that this method opens is to apply certain
filtering logic when specific conditions are met (e.g. take some
decisions on an OS and not on another).

^^^^^^^^^^^^^^^^^^
Validate detection
^^^^^^^^^^^^^^^^^^

To increase detection robustness, packagers may also implement a method
to validate the detected Spec objects:

.. code-block:: python

   @classmethod
   def validate_detected_spec(cls, spec, extra_attributes):
       """Validate a detected spec. Raise an exception if validation fails."""

This method receives a detected spec along with its extra attributes and must

> raise an exception of any type if the spec is considered invalid. In that case

Reviewer 2: Could we restrict this to an instance of "SpackError" and make some subclass available for this?

Author: We certainly could. What would be the gain though? The code that deals with that is:

https://github.com/spack/spack/blob/b52c236a2ca7843042e6049c6e06596aacb2b9aa/lib/spack/spack/cmd/external.py#L279-L288

and I think we would still need to catch and log any `Exception` to make external detection resilient to errors in packages. As I see it right now, we might impose another constraint on users and what we gain is to log the reason why a spec was discarded with a different message. Not sure if this is worth it. Thoughts?

Reviewer 2: > and I think we would still need to catch and log any Exception to make external detection resilient to errors in packages. 

That makes sense to me. I originally thought (based on the function parameters) that users wouldn't be calling filesystem functions etc. and so the only errors raised in this function would be explicitly raised by the user. However, it's true that we don't want an error in a single package to prevent package detection in general.

My request in this case is that users can also return `False` to indicate that the provided spec is not valid (so they don't have to think about what exception to raise).

Author: > My request in this case is that users can also return False to indicate that the provided spec is not valid (so they don't have to think about what exception to raise).

I was implementing this solution, but went back on it as it introduces some asymmetry: users can return `False` or raise, but they are never required to return `True` (and if they are we have a more complex API). I think the concern that users don't have to decide what to raise is a good one so I went with another solution in 150e477.

The idea is that the documentation now says explicitly that users should assert conditions on the spec (thus they don't even have to know that this operation raises an exception). The implementation stays the same as it is now permitting us to catch whatever is going wrong and proceed with the next spec / package. 

Since I went with a different solution I'll leave this discussion open. Let me know if this seems a better approach than before or if you want to revisit it.

Reviewer 2: I think we may want to be prepared for user confusion over this: `assertions` are a slightly different paradigm than if/else/raise. I want to avoid users having to swap between paradigms while implementing packages.

That being said, assertions are OK, but I think it's also worth providing an exception type for users to raise, and to document that in case they want to do things like:

```
if spec_doesnt_meet_my_condition:
  raise NewSpackExceptionIntroducedInThisPR
```

Author: Done in 17b48bf Let me know if that suffices.

the spec will be discarded and the exception message will be logged as the reason
for discarding it.

.. _determine_spec_details:

^^^^^^^^^^^^^^^^^^^^^^^^^
Custom detection workflow
^^^^^^^^^^^^^^^^^^^^^^^^^

In the rare case when the mechanisms described so far don't fit the
detection of a package, the implementation of all the methods above
can be disregarded and instead a custom ``determine_spec_details``
method can be implemented directly in the package class (note that
the definition of the ``executables`` attribute is still required):

.. code-block:: python

   @classmethod
   def determine_spec_details(cls, prefix, exes_in_prefix):
       # exes_in_prefix = a set of paths, each path is an executable
       # prefix = a prefix that is common to each path in exes_in_prefix

       # return None or [] if none of the exes represent an instance of
       # the package. Return one or more Specs for each instance of the
       # package which is thought to be installed in the provided prefix

This method takes as input a set of discovered executables (which match
those specified by the user) as well as a common prefix shared by all

> of those executables. The function must return one or more Specs associated

Reviewer 3: link to Specs docs?

Author: Done

^^^^^^^^^^^^^^^^^
Minimal detection
^^^^^^^^^^^^^^^^^

The first step is fairly simple, as it requires only to
specify a package level ``executables`` attribute:

.. code-block:: python

   class Foo(Package):
       # Each string provided here is treated as a regular expression, and
       # would match for example 'foo', 'foobar', and 'bazfoo'.
       executables = ['foo']

This attribute must be a list of strings. Each string is a regular
expression (e.g. 'gcc' would match 'gcc', 'gcc-8.3', 'my-weird-gcc', etc.) to
determine a set of system executables that might be part or this package. Note
that to match only executables named 'gcc' the regular expression ``r'^gcc$'``
must be used.

Finally to determine the version of each executable the ``determine_version``
method must be implemented:

.. code-block:: python

   @classmethod
   def determine_version(cls, exe):
       """Return either the version of the executable passed as argument
       or ``None`` if the version cannot be determined
       """

This method receives as input the path to a single executable and must return
as output its version as a string, or ``None`` if the version cannot be determined

> or the executable is not related to the package.

Reviewer 2: I suggest rephrasing

> This method receives as input the path to a single executable and must return as output its version as a string, or ``None`` if the version cannot be determined or the executable is not related to the package.

to clarify:

> This method receives as input the path to a single executable and must return as output its version as a string; if the user cannot determine the version or determines that the executable is not an instance of the package, they can return ``None`` and the exe will be discarded as a candidate.

Author: Done in 2320945

> Implementing the three steps above is mandatory, and gives the package the

Reviewer 2: Implementing the *two* steps above

Author: Done in 2320945

basic ability to detect if a spec is present on the system at a given version.

.. note::
   Any executable for which the ``determine_version`` method returns ``None``
   will be discarded and won't appear in later stages of the workflow described below.

^^^^^^^^^^^^^^^^^^^^^^^^
Additional functionality
^^^^^^^^^^^^^^^^^^^^^^^^

> Besides the three mandatory steps described above, there are also optional

Reviewer 2: Besides the *two* mandatory steps described above...

Author: Done in 2320945

1. A variant string
2. A tuple of a variant string and a dictionary of extra attributes
3. A list of items matching either 1. or 2. (if multiple specs are detected
   from the set of executables)

If extra attributes are returned, they will be recorded in ``packages.yaml``
and be available for later reuse. As an example, the ``gcc`` package will record
by default the different compilers found and an entry in ``packages.yaml``
would look like:

.. code-block:: yaml

   packages:
     gcc:
       externals:
       - spec: 'gcc@9.0.1 languages=c,c++,fortran'
         prefix: /usr
         extra_attributes:
           compilers:
             c: /usr/bin/x86_64-linux-gnu-gcc-9
             c++: /usr/bin/x86_64-linux-gnu-g++-9
             fortran: /usr/bin/x86_64-linux-gnu-gfortran-9

This permits, for instance, to keep track of executables that would be named
differently if built by Spack (e.g. ``x86_64-linux-gnu-gcc-9``
instead of just ``gcc``).

.. TODO: we need to gather some more experience on overriding 'prefix'
   and other special keywords in extra attributes, but as soon as we are
   confident that this is the way to go we should document the process.

>    See https://github.com/spack/spack/pull/16526#issuecomment-653783204

Reviewer 2: (question) I wanted to make sure: this PR no longer supports https://github.com/spack/spack/pull/16526#issuecomment-653783204 right? I assumed we would add that in later, but I wanted to check in case you supported it and I missed that.

Author: I added back support for it since it wasn't much effort. Users can add "prefix" or "modules" (or in the future any reserved keyword) to the `extra_attributes` in `determine_variants`. These are popped in:

https://github.com/spack/spack/blob/1d356ca89efd8af0c06aab9ea49e4fcd6d1264d2/lib/spack/spack/package.py#L199-L204

and used to construct the spec. In:

https://github.com/spack/spack/blob/1d356ca89efd8af0c06aab9ea49e4fcd6d1264d2/lib/spack/spack/cmd/external.py#L296-L300

`external_path` is used to set the base_dir if present. I'll add a test to check this case.

Author: Test added in 956c191

Making a package discoverable with ``spack external find``
----------------------------------------------------------

The simplest way to make a package discoverable with
:ref:`spack external find <cmd-spack-external-find>` is to:

1. Define the executables associated with the package
2. Implement a method to determine the versions of these executables

^^^^^^^^^^^^^^^^^
Minimal detection
^^^^^^^^^^^^^^^^^

The first step is fairly simple, as it requires only to
specify a package level ``executables`` attribute:

.. code-block:: python

   class Foo(Package):
       # Each string provided here is treated as a regular expression, and
       # would match for example 'foo', 'foobar', and 'bazfoo'.
       executables = ['foo']

This attribute must be a list of strings. Each string is a regular
expression (e.g. 'gcc' would match 'gcc', 'gcc-8.3', 'my-weird-gcc', etc.) to
determine a set of system executables that might be part or this package. Note

> that to match only executables named 'gcc' the regular expression ``r'^gcc$'``

Reviewer: ```suggestion
that to match only executables named 'gcc' the regular expression ``'^gcc$'``
```
Raw strings are only needed for backslashes, although I'm fine with leaving this in for regexes that might have backslashes.

Author: Done in 450e891

must be used.

Finally to determine the version of each executable the ``determine_version``
method must be implemented:

.. code-block:: python

   @classmethod
   def determine_version(cls, exe):
       """Return either the version of the executable passed as argument

>        or ``None`` if the version cannot be determined

Reviewer: Would be good to add more docs here. It isn't clear if `exe` is a string (name), a string (path), an `Executable` object, or what. Also not clear if version should be a string or a `Version` object.

Author: It's in the text below, but I'll add `Args:` for clarity.

Author: Done in 450e891

       """

This method receives as input the path to a single executable and must return
as output its version as a string; if the user cannot determine the version
or determines that the executable is not an instance of the package, they can
return None and the exe will be discarded as a candidate.
Implementing the two steps above is mandatory, and gives the package the
basic ability to detect if a spec is present on the system at a given version.

.. note::
   Any executable for which the ``determine_version`` method returns ``None``

>    will be discarded and won't appear in later stages of the workflow described below.

Reviewer: 😢

Author: I think it's a sensible request to determine at least the version of a spec. The use cases I have in mind are mainly those of build tools or compilers. The rationale is that most of the times constraints are on versions rather than variants (it's rare to find something that requires `cmake+ncurses` to build but it's not unusual to have requests on a minimum version e.g. `cmake@X.Y:`)

Reviewer: Of course, but it's also not exactly rare for a package to not specify dependency versions. I think this goes back to our discussion of how the new concretizer should handle packages where some versions/variants/etc are unknown. I.e., should it choose the defaults, or should it choose no representation. I would prefer the latter, where `foo+x` requires `foo+x`, `foo~x` requires `foo~x`, and `foo` can depend on either.

^^^^^^^^^^^^^^^^^^^^^^^^
Additional functionality
^^^^^^^^^^^^^^^^^^^^^^^^

Besides the two mandatory steps described above, there are also optional
methods that can be implemented to either increase the amount of details
being detected or improve the robustness of the detection logic in a package.

""""""""""""""""""""""""""""""
Variants and custom attributes
""""""""""""""""""""""""""""""

The ``determine_variants`` method can be optionally implemented in a package
to detect additional details of the spec:

.. code-block:: python

   @classmethod
   def determine_variants(cls, exes, version_str):
       """Return either a variant string, a tuple of a variant string
       and a dictionary of extra attributes that will be recorded in

>        packages.yaml or a list of those items.

Reviewer: Same here, needs more docs on input/outputs.

Author: Done in 450e891

       """

This method takes as input a list of executables that live in the same prefix and
share the same version string, and returns either:

1. A variant string
2. A tuple of a variant string and a dictionary of extra attributes

> 3. A list of items matching either 1. or 2. (if multiple specs are detected

Reviewer: ```suggestion
3. A list of items matching either 1 or 2 (if multiple specs are detected
```

Author: Done in 450e891

   from the set of executables)

If extra attributes are returned, they will be recorded in ``packages.yaml``
and be available for later reuse. As an example, the ``gcc`` package will record
by default the different compilers found and an entry in ``packages.yaml``
would look like:

.. code-block:: yaml

   packages:
     gcc:
       externals:
       - spec: 'gcc@9.0.1 languages=c,c++,fortran'
         prefix: /usr
         extra_attributes:
           compilers:
             c: /usr/bin/x86_64-linux-gnu-gcc-9
             c++: /usr/bin/x86_64-linux-gnu-g++-9
             fortran: /usr/bin/x86_64-linux-gnu-gfortran-9

> This permits, for instance, to keep track of executables that would be named

Reviewer: ```suggestion
This allows us, for instance, to keep track of executables that would be named
```

Author: Done in 450e891

This method receives a detected spec along with its extra attributes and can be
used to check that certain conditions are met by the spec. Packagers can either
use assertions or raise an ``InvalidSpecDetected`` exception when the check fails.
In case the conditions are not honored the spec will be discarded and any message
associated with the assertion or the exception will be logged as the reason for
discarding it.

As an example, a package that wants to check that the ``compilers`` attribute is
in the extra attributes can implement this method like this:

.. code-block:: python

   @classmethod
   def validate_detected_spec(cls, spec, extra_attributes):
       """Check that 'compilers' is in the extra attributes."""
       msg = ('the extra attribute "compilers" must be set for '
              'the detected spec "{0}"'.format(spec))
       assert 'compilers' in extra_attributes, msg

or like this:

.. code-block:: python

   @classmethod
   def validate_detected_spec(cls, spec, extra_attributes):
       """Check that 'compilers' is in the extra attributes."""
       if 'compilers' not in extra_attributes:

>          msg = ('the extra attribute "compilers" must be set for '
>                  'the detected spec "{0}"'.format(spec))
>          raise InvalidSpecDetected(msg)

Reviewer: ```suggestion
           msg = ('the extra attribute "compilers" must be set for '
                   'the detected spec "{0}"'.format(spec))
           raise InvalidSpecDetected(msg)
```

Author: Done in 450e891

To make a package discoverable with
:ref:`spack external find <cmd-spack-external-find>` you must
mark it with the ``detectable`` decorator:

.. code-block:: python

   @detectable
   class Gcc(AutotoolsPackage):
      """The GNU Compiler Collection includes front ends for C, C++, Objective-C,
      Fortran, Ada, and Go, as well as libraries for these languages."""

> define the executables associated with the package and must

Reviewer 2: I'd recommend listing everything up front before providing examples and code samples. Also, I think this should rephrase to emphasize that `determine_version` is required, e.g.:

> To make a package discoverable with `spack external` find, you must mark the Package with the `detectable` decorator and implement a method to determine the version of a package when given an executable for the package.

You can later talk about how additional methods are provided that you can override to get additional details.

implement methods to generate one or more Specs when given
a set of executables.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The ``executables`` attribute
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The executables are specified as a package level ``executables``
attribute which is a list of strings:

.. code-block:: python

   class FooPackage(Package):
       # Each string provided here is treated as a regular expression, and
       # would match for example 'foo', 'foobar', and 'bazfoo'.
       executables = ['foo']

Each string is treated as a regular expression (e.g. 'gcc' would match 'gcc', 'gcc-8.3',
'my-weird-gcc', etc.) to determine a set of system executables that might be part
or this package.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Using the default workflow to detect specs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The detection of system installed packages in Spack is based on

Reviewer 2: My understanding is that minimally a user must implement `determine_version`. IMO it would be good to start with that so the user is introduced to the typical/easier approach.

Author: @scheibelp Docs have been reworked in 6e0f32a to follow the structure suggested here: typical use cases first, advanced/rare use cases later (instead of just describing the mechanics of the current implementation).

calling the ``determine_spec_details`` method for each detectable
package. This method, given a ``prefix`` and a list of matching
``executables``, is responsible of computing the list of
detected Spec objects.

When a package class is marked with the ``detectable`` decorator

> it receives automatically a default implementation of

Reviewer 2: (minor) "receives automatically" -> "automatically receives"

``determine_spec_details`` that splits the overall detection
task into simpler operations, each with its own callback method.

"""""""""""""""""""""""""""
Filter matching executables
"""""""""""""""""""""""""""

> The first method that gets called is:

Reviewer 2: I think mentioning order might be confusing/extraneous - I think this could be rephrased as:

> You can filter the set of executables detected by Spack by implementing  `filter_detected_specs`...

and I think it would be useful to provide an example of when the user might want to do this.

.. code-block:: python

    @classmethod
    def filter_detected_exes(cls, prefix, exes_in_prefix):
        """Return a filtered list of the executables in prefix"""

which takes the prefix and matching executables and returns a
filtered list of said executables. The implementation of this method
is optional and, if none is given, the list of matching executables
is returned unfiltered.

""""""""""""""""""""""""""""""""""""""
Determine the version of an executable
""""""""""""""""""""""""""""""""""""""

> The callback method that follows filtering is:

Reviewer 2: Although it seems repetitive given the title, I think it's good to start each section introducing the concept:

> To make a package detectable, Spack requires you to implement a function to determine the version of a specified executable...

.. code-block:: python

   @classmethod
   def determine_version(cls, exe):
       """Return either the version of the executable passed as argument
       or ``None`` if the version cannot be determined
       """

which receives as input a single executable and must return as output
its version as a string, or ``None`` is the version cannot be determined.
This method must be implemented by a detectable package using the default
workflow.

""""""""""""""""""
Determine variants
""""""""""""""""""

Finally, once versions are determined, a method to determine variants
is called:

.. code-block:: python

   @classmethod
   def determine_variants(cls, exes, version_str):
       """Return either a variant string, a tuple of a variant string
       and a dictionary of extra attributes that will be recorded in
       packages.yaml or a list of those items.
       """

This method takes as input a list of executables that live in the same prefix and
share the same version string, and returns either:

1. A variant string (if a single spec is detected for these executables)

> 2. A tuple of a variant string and a dictionary or extra attributes that need to

Reviewer 2: "and a dictionary __or__ extra attributes" -> "and a dictionary __of__ extra attributes"?

   be recorded in packages.yaml for any reason
3. A list of items matching either 1. or 2. (if multiple specs are detected
   from the set of executables)

> The implementation of this method is optional and if none is given the default

Reviewer 2: I'd move the mention of this being optional to the beginning of this section.

           exe = spack.util.executable.Executable(exe_path)
           output = exe('--version')
           version_str = ...  # parse output for version string
           return Spec.from_detection(
               'foo-package@{0}'.format(version_str)
           )

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Validate detected spec objects
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To increase detection robustness, packagers may also implement a method
to validate the detected Spec objects:

.. code-block:: python

   @classmethod
   def validate_detected_spec(cls, spec, extra_attributes):
       """Validate a detected spec. Raise an exception if validation fails."""

This method receives a detected spec along with its extra attributes and must
raise an exception if the spec is considered invalid. In that case the spec

> will be discarded from the list of the ones detected and the exception message

Reviewer 2: Should this throw a specific exception type?

Also "the spec will be discarded from the list of the ones detected" -> "the spec will be discarded"
